### PR TITLE
[Menubar] Add Menubar primitive

### LINKED ISSUE #1133: [New Primitive] `Menubar`
Following the WAI ARIA pattern documented [here](https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html).

I imagine it would follow a similar approach to `Toolbar` in terms of architecture e.g. provide a `Menubar.Button` that can be composed with `DropdownMenu`. 

Although, do we think that `Menubar` should always be dropdown menus? if the menubar should never contain a button that actions something straight from the bar, perhaps we would enforce that by exposing wrapped dropdown parts as `Menubar.Menu` parts. Thoughts?

Proposals:

```jsx
<Menubar.Root>
  <DropdownMenu.Root>
    <DropdownMenu.Trigger asChild>
      <Menubar.Button />
    </DropdownMenu.Trigger>
    <DropdownMenu.Content>
      <DropdownMenu.Item />
    </DropdownMenu.Content>
  </DropdownMenu.Root>
</Menubar.Root>
```
```jsx
<Menubar.Root>
  <Menubar.Menu>
    <Menubar.Trigger />
    <Menubar.Content>
      <Menubar.Item />
    </Menubar.Content>
  </Menubar.Menu>
</Menubar.Root>
```

https://user-images.githubusercontent.com/175330/152394009-6488db98-3ef1-4ea3-b9ef-126b960997d0.mp4

### PR DESCRIPTION
**Pull request by luisorbaiceta**
This PR is intended for the Menubar implementation. Closes #1133

- [x] Discuss API (Following the [WAI-ARIA pattern for Menubar](https://www.w3.org/WAI/ARIA/apg/example-index/menubar/menubar-editor.html#top))
- [ ] Add integration tests
- [ ] Add stories for visual testing
- [ ] On right or left arrow press, if next focusable element is next tab it should open the tab and focus its trigger as described in WAI-ARIA (depends on #1574)

### [Suggested API](https://github.com/radix-ui/primitives/issues/1133#issue-1123075390)

```tsx
<Menubar.Root>
  <Menubar.Menu>
    <Menubar.Trigger />
    <Menubar.Content>
      <Menubar.Item />
    </Menubar.Content>
  </Menubar.Menu>
  <Menubar.Menu>
    <Menubar.Trigger />
    <Menubar.Content>
      <Menubar.Item />
      <Menubar.SubMenu>
        <Menubar.Trigger />
        <Menubar.Content>
          <Menubar.Item />
        </Menubar.Content>
      </Menubar.SubMenu>
    </Menubar.Content>
  </Menubar.Menu>
</Menubar.Root>
```

For `Menus` and `Submenus` the `Dropdown` implementation will be used. Any additional behaviour will be added via its controlled api

### CODE REVIEW

File: .yarn/versions/6de667dc.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+declined:
+  - primitives
+  - "@radix-ui/react-menu-bar"
```
Comment (jjenzz) [inline .yarn/versions/6de667dc.yml:3]: teehee, I thought the same as you but when I googled the pattern it seems it's actually one word 🙈

`react-menubar`
`const Menubar = React.forwardRef();`

etc.

Comment (luisorbaiceta) [reply .yarn/versions/6de667dc.yml:3]: yep. I noticed that haha. Only had time to change the PR title 😂

Comment (jjenzz) [reply .yarn/versions/6de667dc.yml]: Excited for this! Thanks again 🙂

**Comment by luisorbaiceta** [comment]
There are some tests cases that should go in the `DropdownMenu` and `Menu` primitives. As I haven't seen tests files for those, I'll first put them here till I can find some time to write tests for those components too

**Comment by jjenzz** [comment]
@luisorbaiceta Just a heads up that we favour chromatic or cypress tests these days over jest. I think the plan was to remove our jest tests eventually so I wouldn't invest too much time there.

There is already a `DropdownMenu` cypress test which covers a lot of its requirements but feel free to include more there if needed.

**Comment by luisorbaiceta** [comment]
Good! I'll go with cypress then

**Comment by luisorbaiceta** [comment]
Right now I'm planning to implement the Right/Left arrow navigation by accessing the top level triggers from the `Item` element via the `DropdownMenuItem` controlled API.

There is one extra case that we didn't take into account. When a top level menu trigger is focused, when hitting `ArrowUp`, this should open the menu and focus the last item. I wonder if this should be provided by the DropdownMenu too. Any thoughts?

**Comment by jjenzz** [comment]
> I wonder if this should be provided by the `DropdownMenu` too. Any thoughts?

It does appear to be part of the `DropdownMenu` pattern [here](https://w3c.github.io/aria-practices/examples/menu-button/menu-button-actions.html) so I think it should be provided there but can be done as a separate PR.

**Comment by luisorbaiceta** [comment]
I'll send one tonight, if that's ok 😊

**Comment by jjenzz** [comment]
Just a heads up separately as well, it seems you're going down the route of making the `menubar` for site navigation which is not the intention of the pattern we're going with. Ideally we would follow the example linked to [in the ticket](https://github.com/radix-ui/primitives/issues/1133), so specifically for application menus.

For site navigation we have a `HoverMenu` component in the works which does not use `menubar` / `menu` / `menuitem` roles as they can actually [hinder the user experience](https://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html) when used for site navigation. Most users expect that site navigation can be tabbed through, but `menubar` pattern requires arrow keys.

**Comment by luisorbaiceta** [comment]
I'll leave it up to the user then via the `asChild` prop, and default the Item to the `DropdownMenuItem` current default. Is that ok?

**Comment by jjenzz** [comment]
@luisorbaiceta Spot on 👍  thanks! 😍

**Comment by luisorbaiceta** [comment]
Hi everyone! I've already made a draft for keyboard navigation that could really use some feedback! Thanks in advance! 😊

**Comment by jjenzz** [comment]
@luisorbaiceta Just an FYI that I haven't forgotten about this. We're in the middle of a big push to try get `Toast`, `Select` and `HoverMenu` out the door so focus is on those at the moment but as soon as we've managed to get a release out i'll circle back onto this 🙏

**Comment by luisorbaiceta** [comment]
Hi everyone! Now that `Toast`, `Select` and `HoverMenu` have been released, I would love to keep working on this 😊

**Comment by benoitgrelard** [comment]
Hey @luisorbaiceta,

Sorry for the delay with this. We should have some time now to review and get this going.
Would you mind rebasing your branch and making sure it is up to date with the API of the primitives you are rewrapping?

Thanks!

✌️

**Comment by luisorbaiceta** [comment]
I've noticed some changes in the `Dropdown Menu` components api so I will take some time to adapt everything that has been done so far

**Comment by luisorbaiceta** [comment]
I think I have reached a point in which I could use a review from the team. There are two upstream issues that affect the behavior of this primitive and prevents it from being fully compliant with the WAI-ARIA spec (#1565 #1564 )

Cheers!

**Comment by luisorbaiceta** [comment]
Hi everyone! A review would really help to push this forward 🙃

**Comment by benoitgrelard** [comment]
Hey @luisorbaiceta,

I just wanted to let you know that I have started reviewing this amongst some of the other contributions on the repo.

**Comment by Looky1173** [comment]
Hi, are there any updates on this pull request or the blocking issue #1308?

**Comment by andy-hook** [comment]
Hey @luisorbaiceta , I wanted to thank you personally for all of your hard work and effort on this and other changes to the project.

2022 was a super busy year for us and one of the bigger struggles was finding the time to provide feedback and reviews on this and other changes in a timely manner.  

We did manage to find a gap in our schedule just before Christmas though and we made this work one our key goals to ship in the new year, your original API proposal and code was incredibly helpful in being able to release yesterday, thank you for the effort and patience.

  Closing now as changes implemented in https://github.com/radix-ui/primitives/pull/1846,

**Pull request closed without merging**